lib/srv/discovery/status.go

```diff
@@ -0,0 +1,179 @@
+
+func (d *awsSyncStatus) iterationFinished(fetchers []aws_sync.AWSSync, pushErr error, lastUpdate time.Time) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.awsSyncResults = make(map[string][]awsSyncResult)
+	for _, fetcher := range fetchers {
+		// Only update the status for fetchers that are from the discovery config.
+		if !fetcher.IsFromDiscoveryConfig() {
+			continue
+		}
+
+		count, statusErr := fetcher.Status()
+		statusAndPushErr := trace.NewAggregate(statusErr, pushErr)
+
+		fetcherResult := awsSyncResult{
+			state:               discoveryconfigv1.DiscoveryConfigState_DISCOVERY_CONFIG_STATE_RUNNING.String(),
+			lastSyncTime:        lastUpdate,
+			discoveredResources: count,
+		}
+
+		if statusAndPushErr != nil {
+			errorMessage := statusAndPushErr.Error()
+			fetcherResult.errorMessage = &errorMessage
+			fetcherResult.state = discoveryconfigv1.DiscoveryConfigState_DISCOVERY_CONFIG_STATE_ERROR.String()
+		}
+
+		d.awsSyncResults[fetcher.DiscoveryConfigName()] = append(d.awsSyncResults[fetcher.DiscoveryConfigName()], fetcherResult)
+	}
+}
+
+func (d *awsSyncStatus) discoveryConfigs() []string {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	ret := make([]string, 0, len(d.awsSyncResults))
+	for k := range d.awsSyncResults {
+		ret = append(ret, k)
+	}
+	return ret
+}
+
+func (d *awsSyncStatus) iterationStarted(fetchers []aws_sync.AWSSync, lastUpdate time.Time) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.awsSyncResults = make(map[string][]awsSyncResult)
+	for _, fetcher := range fetchers {
+		// Only update the status for fetchers that are from the discovery config.
+		if !fetcher.IsFromDiscoveryConfig() {
+			continue
+		}
+
+		fetcherResult := awsSyncResult{
+			state:        discoveryconfigv1.DiscoveryConfigState_DISCOVERY_CONFIG_STATE_SYNCING.String(),
+			lastSyncTime: lastUpdate,
+		}
+
+		d.awsSyncResults[fetcher.DiscoveryConfigName()] = append(d.awsSyncResults[fetcher.DiscoveryConfigName()], fetcherResult)
+	}
+}
+
+func mergeErrorMessage(existing *string, currentFetcherMessage *string) *string {
+	if existing == nil {
+		return currentFetcherMessage
+	}
+
+	if currentFetcherMessage == nil {
+		return existing
+	}
+
+	newErrorMessage := *existing + "\n" + *currentFetcherMessage
+	return &newErrorMessage
+}
+
+func (d *awsSyncStatus) mergeIntoGlobalStatus(discoveryConfigName string, existingStatus discoveryconfig.Status) discoveryconfig.Status {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+
+	awsStatusFetchers, found := d.awsSyncResults[discoveryConfigName]
+	if !found {
+		return existingStatus
+	}
+
+	for _, fetcher := range awsStatusFetchers {
+		existingStatus.DiscoveredResources = existingStatus.DiscoveredResources + fetcher.discoveredResources
+
+		// Each DiscoveryConfigStatus has a global State and Error Message, but those are produced per Fetcher.
+		// We choose to keep the most informative states by favoring error states/messages.
+		existingStatus.ErrorMessage = mergeErrorMessage(existingStatus.ErrorMessage, fetcher.errorMessage)
```

Comment:
Merge using a strings.Buffer (or bytes.Buffer) instead of doing multiple, independent string concatenations.

Reply:
Moved to a `strings.Join` which is pretty much the same